### ISSUE TITLE
apache parser searching sites-available, should be sites-enabled

### ISSUE THREAD

**Issue by BillKeenan** [letsencrypt/letsencrypt#1043]
for obvious reasons..

**Comment by jdkasten** [comment]
Right now, the parser searches both... sites-enabled for what configuration directives are actually enabled and sites-available for all of the possible sites that you may have configured, for VirtualHost selection.

My idea was that Apache can automatically enable a site that may have been created, but not yet deployed yet.  If Apache is configured appropriately for Debian, sites-enabled should be a subset of sites-available.

The Apache VirtualHost objects have an attribute for if the site is currently enabled.

I do think that it makes sense to completely "generalize" the configuration by only searching sites-enabled (it would make porting to other layouts easier).

I would suggest that we don't make a distinction between sites-enabled and any other Included directory.  We should simply look through all of the current configuration files and look for VirtualHost blocks.

**Comment by BillKeenan** [comment]
ah, so this is relying on sites-enabled being symlinks to sites-available then?

**Comment by pde** [comment]
@jdkasten so the only thing that would break if we remove sites-available from the search path is enablement of vhosts that the admin has set up but not actually deployed?  In general that sounds slightly dangerous anyway, so if this is a simple fix we should just remove `sites-available` from that search path.

@BillKeenan interested in making a PR?

**Comment by BillKeenan** [comment]
I am. on it.

**Comment by BillKeenan** [comment]
does anyone have some more fullsome docs on getting this running with vagrant.. currently getting this

```
Error: unauthorized :: The client lacks sufficient authorization :: Error creating new authz :: Name is blacklisted

```

after issuing this; sudo ./venv/bin/letsencrypt -d www.example.com auth

**Comment by kuba** [comment]
`www.example.com` is blacklisted, use real domain

**Comment by SebiTimeWaster** [comment]
this prevents me from using letsencrypt, i have no files in sites-available and the letsencrypt client tells me that no config was found and closes.

**Comment by BillKeenan** [comment]
you can use the manual authorization mode until this is resolved., i did, worked well/easy.

**Comment by pde** [comment]
We should definitely fix this!  @BillKeenan are you still interested in writing this PR? If not, we should assign someone the ticket...

**Comment by BillKeenan** [comment]
right! Yes! I have it mostly in the can, I'll get on it tonight.

**Comment by pde** [comment]
@BillKeenan nudge :)

**Comment by BillKeenan** [comment]
hmm, i just ran it, and it's pulling a site from my sites-enabled without issue...

**Comment by pde** [comment]
@BillKeenan do you mean you can no longer reproduce this bug, or that you have a branch that fixes it?  I can definitely still reproduce...

**Comment by BillKeenan** [comment]
Can you provide steps to reproduce?

Bill Keenan
VP Technology
416.583.3772
bill@authenticweb.com

**Comment by pde** [comment]
Sure!  Though I think @joohoi might have started working on this, so it may not be worth spending more time on at the moment. 

```
pde@damoclid:~/letsencrypt/letsencrypt$ ls /etc/apache2/sites-*
/etc/apache2/sites-available:

/etc/apache2/sites-enabled:
001-dee.conf  001-tweedle.conf
pde@damoclid:~/letsencrypt/letsencrypt$ cat /etc/apache2/sites-enabled/*
<VirtualHost *:80>
	ServerAdmin webmaster@localhost
        ServerName dee.isnot.org

	DocumentRoot /var/www/tweedle
	<Directory />
		Options FollowSymLinks
		AllowOverride None
	</Directory>
	<Directory /var/www/>
		Options Indexes FollowSymLinks MultiViews
		AllowOverride None
		Order allow,deny
		allow from all
	</Directory>

	ScriptAlias /cgi-bin/ /usr/lib/cgi-bin/
	<Directory "/usr/lib/cgi-bin">
		AllowOverride None
		Options +ExecCGI -MultiViews +SymLinksIfOwnerMatch
		Order allow,deny
		Allow from all
	</Directory>

	<If "%{REQUEST_URI} =~ /thing/ || %{QUERY_STRING} == 'edit=1'">
		RewriteEngine on
		AuthType basic
		AuthBasicProvider file
		AuthUserFile "/etc/apache2/passwords"
		Require valid-user
	</If>
		

	ErrorLog ${APACHE_LOG_DIR}/error.log

	# Possible values include: debug, info, notice, warn, error, crit,
	# alert, emerg.
	LogLevel warn

	CustomLog ${APACHE_LOG_DIR}/access.log combined
</VirtualHost>
<VirtualHost *:80>
	ServerAdmin webmaster@localhost
        ServerName tweedle.isnot.org

	DocumentRoot /var/www/tweedle
	<Directory />
		Options FollowSymLinks
		AllowOverride None
	</Directory>
	<Directory /var/www/>
		Options Indexes FollowSymLinks MultiViews
		AllowOverride None
		Order allow,deny
		allow from all
	</Directory>

	ScriptAlias /cgi-bin/ /usr/lib/cgi-bin/
	<Directory "/usr/lib/cgi-bin">
		AllowOverride None
		Options +ExecCGI -MultiViews +SymLinksIfOwnerMatch
		Order allow,deny
		Allow from all
	</Directory>

	ErrorLog ${APACHE_LOG_DIR}/error.log

	# Possible values include: debug, info, notice, warn, error, crit,
	# alert, emerg.
	LogLevel warn

	CustomLog ${APACHE_LOG_DIR}/access.log combined
</VirtualHost>
pde@damoclid:~/letsencrypt/letsencrypt$ sudo venv/bin/letsencrypt --apache -t

-------------------------------------------------------------------------------
No names were found in your configuration files.
You should specify ServerNames in your config files in order to allow for
accurate installation of your certificate.
If you do use the default vhost, you may specify the name manually. Would you
like to continue?
-------------------------------------------------------------------------------
(Y)es/(N)o:
```

**Comment by pde** [comment]
Flagging the ticket as joohoi's for the time being.

**Comment by joohoi** [comment]
This should have worked already, but a bug in augeas lib caused the path not to be parsed in some configurations. #2092 fixes that.

**Comment by joohoi** [comment]
More detailed explanation:

This should have worked already - sites-enabled should be parsed because it's "Include":d from apache2.conf in the root directory, and if for some reason it was not, it can be parsed by providing --apache-vhost-root /etc/apache2/sites-enabled in the command line, provided by #1811. 

However in some cases (and especially if defined in the command line) it was NOT being parsed, because of libaugeas bug, that prevents parsing of overlapping match definitions. #2092 fixes that problem, by handling the augeas parsing paths if they overlap. Leaving only the "wildest card" match rule intact

**Comment by joohoi** [comment]
So for now, sites-enabled can be parsed by providing cli parameters --apache-vhost-root /etc/apache2/sites-enabled and --apache-handle-sites False

--apache-vhost-root sets the vhost root to sites-enabled, and --apache-handle-sites disables the a2ensite functionality, as if this is needed, the user doesn't use the sites-available / sites-enabled & a2ensite / a2dissite configuration scheme to begin with.

**Comment by pde** [comment]
This has been partially addressed, but a full fix will have to wait for 0.3.0.
